pkg/csi/service/common/commonco/k8sorchestrator/k8sorchestrator.go

```diff
@@ -18,70 +18,191 @@ package k8sorchestrator
+type K8sVanillaInitParams struct {
+	InternalFeatureStatesConfigInfo config.FeatureStatesConfigInfo
 }
 
 // Newk8sOrchestrator instantiates K8sOrchestrator object and returns this object
-func Newk8sOrchestrator(ctx context.Context, featureStatesConfigMapInfo config.FeatureStatesConfigInfo) (*K8sOrchestrator, error) {
-	var coInstanceErr error
+// NOTE: As Newk8sOrchestrator is created in the init of the driver and syncer components,
+// raise an error only if it is of utmost importance
+func Newk8sOrchestrator(ctx context.Context, controllerClusterFlavor cnstypes.CnsClusterFlavor, params interface{}) (*K8sOrchestrator, error) {
+	var (
+		coInstanceErr error
+		k8sClient     clientset.Interface
+	)
 	onceFork8sOrchestratorInstance.Do(func() {
 		log := logger.GetLogger(ctx)
 		log.Info("Initializing k8sOrchestratorInstance")
 		k8sOrchestratorInstance = &K8sOrchestrator{}
-		k8sOrchestratorInstance.featureStates = make(map[string]string)
-		k8sClient, coInstanceErr := k8s.NewClient(ctx)
+		k8sOrchestratorInstance.clusterFlavor = controllerClusterFlavor
+
+		// Create a K8s client
+		k8sClient, coInstanceErr = k8s.NewClient(ctx)
 		if coInstanceErr != nil {
 			log.Errorf("Creating Kubernetes client failed. Err: %v", coInstanceErr)
 			return
 		}
-		featureStatesConfigMapName = featureStatesConfigMapInfo.Name
-		featureStatesConfigMapNamespace = featureStatesConfigMapInfo.Namespace
-		fssConfigMap, err := k8sClient.CoreV1().ConfigMaps(featureStatesConfigMapNamespace).Get(ctx, featureStatesConfigMapName, metav1.GetOptions{})
-		if err != nil {
-			log.Errorf("failed to fetch configmap %s from namespace %s. Setting the feature states to default values: %v. Error: %v", featureStatesConfigMapName, featureStatesConfigMapNamespace, k8sOrchestratorInstance.featureStates, err)
-		} else {
-			updateFSSValues(ctx, fssConfigMap, k8sOrchestratorInstance)
+
+		coInstanceErr = initOrchestrator(ctx, k8sClient, controllerClusterFlavor, params)
+		if coInstanceErr != nil {
+			log.Errorf("Failed to initialize the orchestrator. Error: %v", coInstanceErr)
+			return
 		}
-		// Set up kubernetes resource listeners for metadata syncer
-		k8sOrchestratorInstance.informerManager = k8s.NewInformer(k8sClient)
-		k8sOrchestratorInstance.informerManager.AddConfigMapListener(
-			func(obj interface{}) {
-				configMapAdded(obj, k8sOrchestratorInstance)
-			}, // Add
-			func(oldObj interface{}, newObj interface{}) { // Update
-				configMapUpdated(oldObj, newObj, k8sOrchestratorInstance)
-			},
-			func(obj interface{}) {
-				configMapDeleted(obj, k8sOrchestratorInstance)
-			}) // Delete
+
 		log.Info("k8sOrchestratorInstance initialized")
-		k8sOrchestratorInstance.informerManager.Listen()
 	})
 	return k8sOrchestratorInstance, coInstanceErr
 }
 
+// initOrchestrator initializes the orchestrator parameters
+func initOrchestrator(ctx context.Context, k8sClient clientset.Interface, controllerClusterFlavor cnstypes.CnsClusterFlavor, params interface{}) error {
```

Comment:
My initial intention was to move out all the init code to a different function so that the `once.Do` part is clear and concise. Any more additions to the init of k8s orchestrator could be added here. But I will remove this function now that we are changing the `once.Do` part.

Reply:
Removed this function in the latest commit